ping/ping_output.c

```diff
@@ -0,0 +1,255 @@
+{
+	if (!rts->opt_json)
+		printf(_(" parse error (%s)"), reason);
+	construct_json(rts, PING_JSON_STR, "error", reason);
+}
+
+inline void ping_print_error_qtype(struct ping_rts *rts, uint32_t val)
+{
+	if (!rts->opt_json)
+		printf(_(" unknown qtype(0x%02x)"), val);
+	construct_json(rts, PING_JSON_STR, "error", "unknown qtype");
+}
+
+inline void ping_print_truncated(struct ping_rts *rts)
+{
+	if (rts->opt_json) {
+		construct_json(rts, PING_JSON_STR, "info", "truncated");
+		return;
+	}
+
+	printf(_(" (truncated)"));
+	if (rts->ni.subject_type == IPUTILS_NI_ICMP6_SUBJ_IPV4)
+		putchar('\n');
+}
+
+inline void ping_print_packet(struct ping_rts *rts)
+{
+	if (!rts->opt_json) {
+		char *source;
+		char *target;
+
+		if (rts->ni.subject_type == IPUTILS_NI_ICMP6_SUBJ_IPV6) {
+			source = pr_addr(rts, &rts->source6, sizeof(rts->source6));
+			target = pr_raw_addr(rts, &rts->whereto6, sizeof(rts->whereto6));
+		} else {
+			source = inet_ntoa(rts->source.sin_addr);
+			target = inet_ntoa(rts->whereto.sin_addr);
+		}
+
+		printf(_("PING %s (%s) "), rts->hostname, target);
+
+		if (rts->ni.subject_type == IPUTILS_NI_ICMP6_SUBJ_IPV6 && rts->flowlabel)
+			printf(_(", flow 0x%05x, "), (unsigned int)ntohl(rts->flowlabel));
+
+		if (rts->device || rts->opt_strictsource) {
+			int saved_opt_numeric = rts->opt_numeric;
+
+			rts->opt_numeric = 1;
+			rts->opt_numeric = saved_opt_numeric;
+
+			printf(_("from %s %s: "), source, rts->device ? rts->device : "");
+		}
+
+		if (rts->ni.subject_type == IPUTILS_NI_ICMP6_SUBJ_IPV6)
+			printf(_("%d data bytes\n"), rts->datalen);
+		else
+			printf(_("%d(%d) bytes of data.\n"), rts->datalen, rts->datalen + 8 + rts->optlen + 20);
+	}
+}
+
+static long llsqrt(long long a)
+{
+	long long prev = LLONG_MAX;
+	long long x = a;
+
+	if (x > 0) {
+		while (x < prev) {
+			prev = x;
+			x = (x + (a / x)) / 2;
+		}
+	}
+
+	return (long)x;
+}
+
+inline void ping_print_statistics(struct ping_rts *rts, struct timespec tv)
+{
+	double tmdev;
+	char rttmin[30];
+	char rttavg[30];
+	char rttmax[30];
+	char rttmdev[30];
+	long long tmvar;
+
+	long total = rts->nreceived + rts->nrepeats;
+	long tmavg = rts->tsum / total;
+
+	if (rts->tsum < INT_MAX)
+		/* This slightly clumsy computation order is important to avoid
+		 * integer rounding errors for small ping times.
+		 */
+		tmvar = (rts->tsum2 - ((rts->tsum * rts->tsum) / total)) / total;
+	else
+		tmvar = (rts->tsum2 / total) - (tmavg * tmavg);
+
+	tmdev = llsqrt(tmvar);
+
+	sprintf(rttmin, "%ld.%03ld", (long)rts->tmin / 1000, (long)rts->tmin % 1000);
+	sprintf(rttavg, "%lu.%03ld", (unsigned long)(tmavg / 1000), (long)(tmavg % 1000));
+	sprintf(rttmax, "%ld.%03ld", (long)rts->tmax / 1000, (long)rts->tmax % 1000);
+	sprintf(rttmdev, "%ld.%03ld", (long)tmdev / 1000, (long)tmdev % 1000);
+
+	if (!rts->opt_json)
+		printf(_("rtt min/avg/max/mdev = %s/%s/%s/%s ms"),
+			rttmin, rttavg, rttmax, rttmdev);
+
+	construct_json_statistics(rts, tv, rttmin, rttavg, rttmax, rttmdev);
```

Comment:
This part is a bit strange, we shouldn't call to the `construct_json_statistics()` unconditionally just to print a newline from that functions `if (!rts->opt_json)`. Rather than that this should really be:

```
if (!rts->opt_json) {
		printf(_("rtt min/avg/max/mdev = %s/%s/%s/%s ms"),
			     rttmin, rttavg, rttmax, rttmdev);
        putchar('\n');
} else {
        construct_json_statistics(rts, tv, rttmin, rttavg, rttmax, rttmdev);
}
```

Generally we shouldn't call to the functions that build and print json if json output is not enabled.

Reply:
Good point, adjusted.